### DISCUSSION TITLE
CBR vs VBR Loryta (and maybe ROI)?

### DISCUSSION THREAD

**Discussion by rhatguy** [blakeblackshear/frigate#3299] [General]
I recently bought a set of Loryta 5442 cams as per Blake's suggestion in the docs.  I set them up as per the doc site also (https://docs.frigate.video/guides/camera_setup#example-camera-configuration).  When picking the bit rate type for the two substreams, I chose VBR, thinking why not use less bandwidth/processing power on my frigate machine.  However I started noticing that the RTMP and the detect streams looked fairly blockey sometimes.  It looks like the bit rate was to low.  Upon bringing the streams up in vlc and navigating to the streams statistics, I noticed that both streams were frequently falling well below (sometimes down into the 300kb range) my max bitrate which I had set a 1536kb.  I also noticed that if I set the bit rate type to CBR the image immediately stabilized and the bit rate fluctuated significantly less.  I did have the quality set to 6 anticipating that to force VBR to get the best quality it could UP TO my max bit rate.

Do others run VBR or CBR streams for your lower resolution streams?  @blakeblackshear , I would if we could get a screenshot of your Loryta/Dahua setup for your cams showing the entire substream configuration for comparison.  That might be easier than listing the text in the docs.

Secondarily, has anyone played with ROI (Region of interest) on the Loryta/Dahua cams?  I'm wondering if it would be beneficial to exclude certain areas like tree branches that frequently blow in the wind from my ROI to avoid wasting compression bandwidth on them.

**Comment by NickM-27** [comment]
I don't use these cameras but I believe CBR is recommended in general and that is what I use with my amcrest cameras

**Reply by esmoyer** [reply]
> my amcrest cameras

I'd be curious what your video settings on your Amcrests are. I'm interested in learning how to optimize my setup but don't know enough to understand the "why" to change something and would be nice to compare. I am also only able to get a resolution of 704x480 on the sub-stream. Is that what you use for the detect stream?

**Reply by NickM-27** [reply]
So naturally it is a very "it depends" situation. If a camera is implemented perfectly then CBR and VBR should behave as well for frigates purposes. The problem is with VBR the camera will compress images more to save bandwidth on the stream. This can lead to smearing or in general less information (as well as high variance in amount of storage used). 

CBR is recommended for higher likelyhood of a smooth experience, but feel free to try VBR and see if things work out. I imagine a nice camera like that it should work as well.

I used to use the 956 x 720 (or whatever) substream of my amcrest doorbell and the VGA stream of my backyard camera. However once I installed doubletake they simply were not high resolution enough to get accurate face detection. Since then I am using the full resolution stream for detect and setting the resolution (`detect -> height` & `detect -> width`) to half (1080p) of the full size stream. This way full resolution is still used for recordings and I get a high quality stream (without being too ridiculous) for detection.

This isn't recommended though since it takes considerably more CPU resources to do the resizing constantly like that, but my server that runs frigate is able to handle it smoothly as it is quite capable.

The other important settings to look at is making sure i-frames interval is the same as the fps (if possible).

**Reply by esmoyer** [reply]
Ok, that makes sense and is helpful. On an semi-unrelated note, you made me double-check and realize my detect height and width were different than what my cameras were set to and was causing some high CPU usage, so thanks for that too!

**Reply by NickM-27** [reply]
Yep, it does have quite the impact

**Reply by dimatx** [reply]
@NickM-27 For your setup where you use your HQ stream for detect, and downsize it with your GPU to half of the full size stream.... What is the FPS of your stream? I assume it's between 15-30? Is that too high to feed into detect? I have 5 TM5442AS cameras and two corals. Run on an Intel NUC <11th gen (has VAAPI). I found the archival quality is better if the iframe interval is FPS*4, but from the guidance it seems like it's best to have it match the FPS for optimal detection.

**Reply by NickM-27** [reply]
There's a lot of things in your question which seem to be conflating entirely different subjects 

> @NickM-27 For your setup where you use your HQ stream for detect, and downsize it with your GPU to half of the full size stream.... What is the FPS of your stream? I assume it's between 15-30?

I have it configured for 10 fps with an iframe interval of 20 (but that is set due to substream live view start times, not for detect quality reasons 

> Is that too high to feed into detect? 

Your detect fps doesn't need to match the camera, you can set it to whatever you want and it will drop unused frames. 

> I have 5 TM5442AS cameras and two corals. Run on an Intel NUC <11th gen (has VAAPI). I found the archival quality is better if the iframe interval is FPS*4, but from the guidance it seems like it's best to have it match the FPS for optimal detection.

Detection is unrelated to iframe interval as far as I know. You want your camera fps and detect fps to match so you're not wasting resources decoding and then discarding frames from the camera. 

That being said, on my cameras I have all the streams set to have an iframe every second.

**Reply by dimatx** [reply]
Thanks for clarifying Nick, and thanks for your patience with me muddling some (or all 😉) of the concepts.

From reading a bit online and some experimentation, it seemed like having a key frame that's 2-4x the FPS resulted in a higher quality stream from an archival POV. Perhaps that's a misguided conclusion and something else was at play there.

**RE: detection + fps:** Is my understanding correct in that having a high(er) FPS on the incoming detect stream than the detect fps setting in Frigate should _not_ result in significant overhead in processing by CPU/GPU? In other words, if I were to feed a 1080p stream at 30fps for detect (so I can reuse that stream for other purposes), and had my detect set to 5 FPS, that alone shouldn't result in lots of wasted resources.

**RE: recording stream + FPS:** My assumption is that, storage availability/cost aside, one might as well use high(er) FPS video (say 30 fps) for archival purposes with the maximum resolution stream the camera provides. Are there tradeoffs to consider here that I may not be taking into account?

**Reply by NickM-27** [reply]
> From reading a bit online and some experimentation, it seemed like having a key frame that's 2-4x the FPS resulted in a higher quality stream from an archival POV. Perhaps that's a misguided conclusion and something else was at play there.

I mean, the higher the iframe interval the higher the quality (because it just means there are more frequent full frames vs motion only frames). That being said, for recording you may not need that many iframes since it will use more bandwidth and won't necessarily have a benefit. Perhaps that is what you mean by "archival POV"


> **RE: detection + fps:** Is my understanding correct in that having a high(er) FPS on the incoming detect stream than the detect fps setting in Frigate should _not_ result in significant overhead in processing by CPU/GPU? In other words, if I were to feed a 1080p stream at 30fps for detect (so I can reuse that stream for other purposes), and had my detect set to 5 FPS, that alone shouldn't result in lots of wasted resources.

Obviously what qualifies as "lots" is subjective, but I would say comparing the scenarios where the stream itself is set to 30 vs 5 fps you would be talking a few % extra on the CPU and GPU

> **RE: recording stream + FPS:** My assumption is that, storage availability/cost aside, one might as well use high(er) FPS video (say 30 fps) for archival purposes with the maximum resolution stream the camera provides. Are there tradeoffs to consider here that I may not be taking into account?

Seems fine, but my recommendation aligns with that of ipcamtalk.com which is that if you actually look at different fps compared, I see basically no benefit past 20fps

**Reply by kirsch33** [reply]
> There's a lot of things in your question which seem to be conflating entirely different subjects
> 
> > @NickM-27 For your setup where you use your HQ stream for detect, and downsize it with your GPU to half of the full size stream.... What is the FPS of your stream? I assume it's between 15-30?
> 
> I have it configured for 10 fps with an iframe interval of 20 (but that is set due to substream live view start times, not for detect quality reasons
> 
> > Is that too high to feed into detect?
> 
> Your detect fps doesn't need to match the camera, you can set it to whatever you want and it will drop unused frames.
> 
> > I have 5 TM5442AS cameras and two corals. Run on an Intel NUC <11th gen (has VAAPI). I found the archival quality is better if the iframe interval is FPS*4, but from the guidance it seems like it's best to have it match the FPS for optimal detection.
> 
> Detection is unrelated to iframe interval as far as I know. You want your camera fps and detect fps to match so you're not wasting resources decoding and then discarding frames from the camera.
> 
> That being said, on my cameras I have all the streams set to have an iframe every second.

im confused, you mention that you have a 10fps camera with iframe set to 20, so thats 1 iframe every 2 seconds, not every second like your last line says? I might be misunderstanding

**Reply by NickM-27** [reply]
> im confused, you mention that you have a 10fps camera with iframe set to 20, so thats 1 iframe every 2 seconds, not every second like your last line says? I might be misunderstanding

all my main streams have their settings configured to have an iframe every second. However, my streams that are being resized and encoded in go2rtc have an iframe every 2 seconds because it is only used for detect and also an iframe every second would be much more resource intensive

**Comment by moandcompany** [comment]
The Lortya 5442 is a great camera. It's a rebranded Dahua 5442 and accepts Dahua OEM firmwares.

I've found that CBR is generally preferable to VBR in most situations where your priorities are to have clear video and extracted frames for subject or object identification. VBR provide some bandwidth and storage savings, but at a usability tradeoff.

You can use the camera's `region of interest` feature to de-emphasize certain regions of the frame and reduce video bandwidth. If you want to take that further, you can also use `privacy masks` to further reduce video bandwidth; this setting can be found in the `overlay` section of your camera video settings.

The [IPCamTalk forum has a good discussion of the CBR vs VBR topic using the Dahua 5442](https://ipcamtalk.com/threads/dahua-cams-vbr-vs-cbr-would-you-expect-a-higher-amount-of-motion-blur.46000/).

Related:  It's also worth looking at the following items to dial-in your camera settings:
- Frame rate:  Assuming a fixed target bitrate and resolution, lower frame rate settings will allow for more information to be recorded per frame, than a higher bitrate (e.g. a 15fps video will have more information per frame than a 30fps video at 1Mbps)
- Exposure settings:  Shutter speed will have the greatest effect on your video and images when it comes to motion blur; make sure your exposure settings specify a sufficiently high minimum-shutter speed to capture moving subject details. This will very relevant in low-light/nighttime conditions.

**Reply by doctorttt** [reply]
I prefer CBR for recording but I believe frigate defaults CBR at like 8100kbps which is high in my opinion. Would be greate to adjust CBR to 5000kbps.

**Reply by moandcompany** [reply]
AFAIK, the bitrate for Frigate recordings is based on the camera's video stream configuration as it writes a copy of the stream from the IP camera and there is not re-encoding or transcoding.

**Reply by NickM-27** [reply]
Yes that's correct, frigate does not do any encoding itself so whatever the bit rate is it is set by you the user
